### PR TITLE
Bug 1903821: contention: Contend over route ingress admission condition 

### PR DESCRIPTION
**Pull request by sgreene570**
**contention: Contend over route ingress admission condition **

pkg/router/controller/contention.go: Enhance `ingressEqual` to compare
the given route ingress' admission condition, ignoring transition
times. This is to remedy situations where router replicas disagree
explicitly on an ingress admission condition for a route when no changes
to a route's spec or status (excluding status.conditions) are made,
such as when an ingress controller switches from a `namespaceOwnership`
policy of `InterNamespaceAllowed` to `Strict`.

vendor: pull in go-cmp/cmp/cmpopts source.
(go-cmp is already in go.mod)

### CODE REVIEW

**Comment by sgreene570** [comment]
@Miciah @frobware are there any fundamental issues with this approach? Please TAL when you can.

File: pkg/router/controller/contention.go [hunk 1]
```diff
@@ -243,7 +246,25 @@ func (t *SimpleContentionTracker) Clear(id string, current *routev1.RouteIngress
 }
 
 func ingressEqual(a, b *routev1.RouteIngress) bool {
-	return a.Host == b.Host && a.RouterCanonicalHostname == b.RouterCanonicalHostname && a.WildcardPolicy == b.WildcardPolicy && a.RouterName == b.RouterName
+	return a.Host == b.Host && a.RouterCanonicalHostname == b.RouterCanonicalHostname && a.WildcardPolicy == b.WildcardPolicy && a.RouterName == b.RouterName &&
+		cmp.Equal(findMostRecentIngressCondition(a), findMostRecentIngressCondition(b), cmpopts.IgnoreFields(routev1.RouteIngressCondition{}, "LastTransitionTime"))
```
Comment (Miciah) [inline pkg/router/controller/contention.go:250]: Would checking the `Admitted` condition be sufficient?  It seems simpler and easier to reason about.  
```suggestion
		cmp.Equal(findCondition(a, routev1.RouteAdmitted), findCondition(b, routev1.RouteAdmitted), cmpopts.IgnoreFields(routev1.RouteIngressCondition{}, "LastTransitionTime"))
```

Comment (sgreene570) [reply pkg/router/controller/contention.go]: Yes, that's definitely the preferable approach (for some reason I didn't realize that a route ingress can only have one _admission_ condition). Fixed, thanks!

**Comment by sgreene570** [comment]
Need to verify that the newest changes still resolve the bug (they should, I just want to double check so the BZ doesn't fail QA).
/hold

**Comment by sgreene570** [comment]
Launcher a cluster-bot cluster with this PR and the new changes look good!
/hold cancel
@Miciah PTAL when you get the chance.

**Comment by Miciah** [comment]
Sorry for the delay!
/lgtm

**Comment by sgreene570** [comment]
> Sorry for the delay!
> /lgtm

Thanks! :grin:

**Comment by Miciah** [comment]
#233 merged in time to ship in OpenShift, but #240 did not.  That could cause confusion but shouldn't cause additional problems (other than that the original problem isn't completely fixed in 4.7).

File: pkg/router/controller/contention.go [hunk 2]
```diff
@@ -243,7 +246,10 @@ func (t *SimpleContentionTracker) Clear(id string, current *routev1.RouteIngress
 }
 
 func ingressEqual(a, b *routev1.RouteIngress) bool {
-	return a.Host == b.Host && a.RouterCanonicalHostname == b.RouterCanonicalHostname && a.WildcardPolicy == b.WildcardPolicy && a.RouterName == b.RouterName
+	// In addition to the RouteIngress' string fields, compare the available admission condition to determine
+	// if the given ingress' are equal. See https://bugzilla.redhat.com/show_bug.cgi?id=1908389.
```
Comment (Miciah) [inline pkg/router/controller/contention.go:250]: Whoops, this was presumably supposed to be https://bugzilla.redhat.com/show_bug.cgi?id=1903821.

**Pull request merged**